### PR TITLE
DAOS-14347 control: Enable setting of port numbers in config generate

### PR DESCRIPTION
**Pull request by tanabarr**
Allow config generate to produce output that contains custom port
numbers for both the control server and for fabric interfaces rather
than just using hardcoded values.  This new behaviour will work in
both dmg and daos_server config generate commands.

Supplying a custom port number to all of the comma-separated access
point addresses in the commandline option will set the (control server)
port parameter in the output server config file
e.g. --access-points=bob:10002,jeff:10002,ben:10002.

Fabric interface port numbers can be set by specifying a comma
separated list in a new commandline option. There must be enough
values to supply all generated engines e.g. --fabric-ports=12460,13460.

Also:
- Mock API backend call to improve dmg unit test coverage.
- Reduce excessive parameters by passing request.
- Only ERROR and more severe log levels are logged to stderr during
  config generate command invocation unless --debug option specified.
- Allow tmpfs non-MD-on-SSD mode (ephemeral) config generation by
  keying MD-on-SSD mode off the control_metadata cli option.
- Update docs to reflect changes.

Features: control
Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/lib/control/auto.go
Comment (kjacque) [inline src/control/lib/control/auto.go:115]: Do these need to be verified similarly to the other parameters below?

File: src/control/lib/control/auto.go
Comment (kjacque) [inline src/control/lib/control/auto.go:1038]: I think we should accept the case where the user passes in a single port that will be used for both engines, if the engines are on different interfaces.

**Review by kjacque — CHANGES_REQUESTED** [review]

**Review by knard-intel — COMMENTED** [review]
Mostly OK for me.
Just have the same concerns as @kjacque regarding the needs to have different ports if the interface are not the same.

File: src/control/lib/control/auto.go
Comment (tanabarr) [reply src/control/lib/control/auto.go:1038]: apparently this should not be allowed according to @Michael-Hennecke

Comment (Michael-Hennecke) [reply src/control/lib/control/auto.go:1038]: we have seen failures many times when the port numbers of the two engine interfaces are identical, even when the fabric interfaces are different. i do not understand the details _why_ that is, but i can attest _that_ it happens. 

and more pressing than the changes here would be a clear guidance how big the _range_ of ports used by an engine actually is. saying "A difference of ~1000 is recommended." isn't really a satisfactory answer because we do not explain how we arrive at that estimate. on which other config parameters or cluster size metrics does this depend? rather than a single port number, i would actually prefer an interval of port numbers in the fabric_iface_port statement, and then the code can check how many ports it actually needs and fail if the interval is too small.

Comment (tanabarr) [reply src/control/lib/control/auto.go:1038]: I have requested clarification on this guidance and the rational behind it from @frostedcmos . The ask for actual port number validation and interval range in the config is nontrivial and the work would require scheduling so please create a ticket if you want this feature in 2.6 (or another) release @Michael-Hennecke . In terms of what changes you require to this PR specifically is it an expansion of the port number difference statement to include some detail on how to calculate the interval? or should I just remove that statement for the moment and follow up with a subsequent PR after I have found some more definitive explanation?

File: src/control/server/config/faults.go
Comment (tanabarr) [reply src/control/server/config/faults.go:125]: apparently not as per @Michael-Hennecke 's guidance

File: utils/config/daos_server.yml
Comment (tanabarr) [reply utils/config/daos_server.yml:354]: Apparently the port numbers need to be unique even if on different interfaces so I won't change that wording. I have simplified the comment  and with the pool specific information to directly above the port definition as suggested.

File: utils/config/daos_server.yml
Comment (tanabarr) [reply utils/config/daos_server.yml:527]: as above

File: src/control/lib/control/auto.go
Comment (Michael-Hennecke) [reply src/control/lib/control/auto.go:1038]: no change requested from me as this point - yes let's move the "interval" topic to a separate ticket.

**Review by Michael-Hennecke — APPROVED** [review]

**Review by knard-intel — DISMISSED** [review]
Thanks for the clarification of the Interface and port numbers

File: src/control/lib/control/auto.go
Comment (tanabarr) [reply src/control/lib/control/auto.go:1038]: I've reverted port validation changes in this PR, @Michael-Hennecke is going to discuss the issue of identical port numbers across engines with different interfaces with @frostedcmos and we will address those changes separately.

Comment (kjacque) [reply src/control/lib/control/auto.go:1038]: OK with me. Thanks for the explanation.

**Review by kjacque — APPROVED** [review]

**Review by knard-intel — APPROVED** [review]
LGTM

File: docs/admin/deployment.md
Comment (mjmac) [inline docs/admin/deployment.md:242]: Why does this flag enable MD on SSD? I would think that should be keyed off of the `--control-metadata-path` flag instead, as that's what the server logic does.

Comment (tanabarr) [reply docs/admin/deployment.md:242]: I could make that change, currently the behaviour is as described. Would you like to do that in this PR?

Comment (mjmac) [reply docs/admin/deployment.md:242]: @kjacque: Is my understanding correct re: the server logic here? If we've got an inconsistency then it seems like it should be resolved...

Comment (tanabarr) [reply docs/admin/deployment.md:242]: what's the inconsistency? the config generate flag is explicit and indicates that the user wants to generate a tmpfs config rather than use any present pmem. IMO that's valid and in this case explicit is better than implicit. I'm happy to change if the consensus is strong that we should automate that selection based on the presence of another option but I'm not seeing any obvious inconsistency.

Comment (mjmac) [reply docs/admin/deployment.md:242]: > what's the inconsistency?

If I am understanding correctly, based on the documentation here, using the `--use-tmpfs-scm` flag enables MD-on-SSD mode. Whereas when the configuration is generated manually, MD-on-SSD mode is only enabled when the `control_metadata:` parameter is set. I'm just saying that it's inconsistent (and confusing) to have two different ways of enabling MD-on-SSD. Furthermore, doing it this way would mean that there's no way to use tmpfs for emulated pmem in non-MD-on-SSD mode anymore, at least via generated configuration.

Comment (tanabarr) [reply docs/admin/deployment.md:242]: ok, so I will remove the --use-tmpfs-scm flag and enable tmpfs-MD-on-SSD config generation when control_metadata commandline option is supplied. does that sound like what you want?

Comment (mjmac) [reply docs/admin/deployment.md:242]: I'm not objecting to the `--use-tmpfs-scm` flag itself. I think there's still a use case for emulating pmem in non-MD-on-SSD (aka legacy) mode. I just wanted to point out that the code as implemented currently uses the presence of a populated `control_metadata:` section in the server config to toggle between the modes. So it didn't make sense to me to tie the mode toggle to the use of tmpfs or not.

Comment (tanabarr) [reply docs/admin/deployment.md:242]: I will update to allow for ephemeral config and key MD-on-SSD off control_metadata option.

Comment (tanabarr) [reply docs/admin/deployment.md]: done, please could you review @mjmac

File: src/control/server/storage/config.go
Comment (knard-intel) [inline src/control/server/storage/config.go:457]: Nit
```suggestion
		return errors.New("external metadata path invalid with DCPM scm class")
```

File: docs/admin/deployment.md
Comment (knard-intel) [inline docs/admin/deployment.md:381]: Nit, a minimal value or a rough value of delay could be helpful for the end user.

**Review by knard-intel — APPROVED** [review]
LGTM, just some minor remarks if you have to repush

File: docs/admin/deployment.md
Comment (tanabarr) [reply docs/admin/deployment.md]: fixed

File: src/control/server/storage/config.go
Comment (tanabarr) [reply src/control/server/storage/config.go]: fixed

**Comment by tanabarr** [comment]
manual verification of PR:
[x] supply custom port numbers for 2 fabric interfaces
[x] create pmem/dcpm config
[x] create tmpfs/ram config (-t option)
[x] create md-on-ssd config (-t and -m options)
[x] perform all the above with daos_server and dmg tools

**Review by knard-intel — APPROVED** [review]

File: src/control/cmd/daos_server/auto_test.go
Comment (kjacque) [inline src/control/cmd/daos_server/auto_test.go:239]: mdonssd should be bumpy case: mdOnSSD

**Review by kjacque — APPROVED** [review]
New changes LGTM. One tiny comment, not needed to land.

File: src/control/cmd/daos_server/auto_test.go
Comment (tanabarr) [reply src/control/cmd/daos_server/auto_test.go:239]: will address if I have to rePush

**Comment by tanabarr** [comment]
gatekeeper, has 2 approvals but don't know why Kris's green tick isn't appearing in the top right of the page

Codespell and pylint issues are related to code unchanged in this PR

File: docs/admin/deployment.md
Comment (ashleypittman) [inline docs/admin/deployment.md:126]:
```suggestion
  -j, --json                                Enable JSON output
```

File: src/control/cmd/daos_server/auto_test.go
Comment (ashleypittman) [reply src/control/cmd/daos_server/auto_test.go:239]: What's the difference between bumpycase and camelcase?

Comment (tanabarr) [reply src/control/cmd/daos_server/auto_test.go:239]: they are the same

File: docs/admin/deployment.md
Comment (tanabarr) [reply docs/admin/deployment.md:126]: Will fix in a follow-up PR

File: src/control/cmd/daos_server/auto_test.go
Comment (tanabarr) [reply src/control/cmd/daos_server/auto_test.go:239]: Will fix in a follow-up PR

**Pull request merged**
